samples/triangle/main.c

```diff
@@ -0,0 +1,281 @@
+        now = XGetTickCount();
+        if ((now-last) > 1000) {
+            fps = frames;
+            frames = 0;
+            last = now;
+        }
+
+        /* Exit the demo after timeout */
+        if ((reboot_after > 0) && ((now-start) > reboot_after)) {
+            break;
+        }
+    }
+
+    MmFreeContiguousMemory(alloc_vertices);
+    pb_show_debug_screen();
+    pb_kill();
+    HalReturnToFirmware(ReturnFirmwareQuickReboot);
+}
+
+/* Construct a viewport transformation matrix */
+static void matrix_viewport(float out[4][4], float x, float y, float width, float height, float z_min, float z_max)
+{
+    memset(out, 0, 4*4*sizeof(float));
+    out[0][0] = width/2.0f;
+    out[1][1] = height/-2.0f;
+    out[2][2] = z_max - z_min;
+    out[3][3] = 1.0f;
+    out[3][0] = x + width/2.0f;
+    out[3][1] = y + height/2.0f;
+    out[3][2] = z_min;
+}
+
+/* Load the shader we will render with */
+static void init_shader(void)
+{
+    uint32_t *p;
+    int       i;
+
+    /* Setup vertex shader */
+    uint32_t vs_program[] = {
+        #include "vs.inl"
+    };
+
+    p = pb_begin();
+
+    /* Set run address of shader */
+    pb_push(p++, NV097_SET_TRANSFORM_PROGRAM_START, 1); *(p++)=0;
+
+    /* Set execution mode */
+    pb_push(p++, NV097_SET_TRANSFORM_EXECUTION_MODE, 2);
+    *(p++) = SHADER_TYPE_EXTERNAL;
+    *(p++) = SHADER_SUBTYPE_REGULAR;
+
+    /* Set cursor and begin copying program */
+    pb_push(p++, NV097_SET_TRANSFORM_PROGRAM_LOAD, 1); *(p++)=0;
+    for (i=0; i<sizeof(vs_program)/8; i++) {
+        pb_push(p++,NV097_SET_TRANSFORM_PROGRAM,4);
+        memcpy(p, &vs_program[i*4], 4*4);
+        p+=4;
+    }
+    pb_end(p);
+
+    /* Setup fragment shader */
+    p = pb_begin();
+    #include "ps.inl"
+    pb_end(p);
+}
+
+/* Set an attribute pointer */
+static void set_attrib_pointer(unsigned int index, unsigned int format, unsigned int size, unsigned int stride, const void* data)
+{
+    uint32_t *p = pb_begin();
+    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_FORMAT + index*4,
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_TYPE, format) | \
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_SIZE, size) | \
+        MASK(NV097_SET_VERTEX_DATA_ARRAY_FORMAT_STRIDE, stride));
+    p += 2;
+    pb_push1(p, NV097_SET_VERTEX_DATA_ARRAY_OFFSET + index*4, (uint32_t)data & 0x03ffffff);
+    p += 2;
+    pb_end(p);
+}
+
+/* Send draw commands for the triangles */
+static void draw_arrays(unsigned int mode, int start, int count)
+{
+    static int overflow_estimation=0;
+    #define MAX_VERTICES_PER_BATCH 65535
+    int num_batches=(count+MAX_VERTICES_PER_BATCH-1)/MAX_VERTICES_PER_BATCH;
+    uint32_t offset=0;
+    while(num_batches) {
+        int n;
+        if (num_batches > 1) {
+            n = MAX_VERTICES_PER_BATCH;
+        } else {
+            n = count % MAX_VERTICES_PER_BATCH;
+        }
+
+        uint32_t *p = pb_begin();
+        pb_push1(p, NV097_SET_BEGIN_END, mode); p += 2;
+
+        uint32_t size=(n+251)/252;
```

Comment:
I think it tries to avoid buffer overruns in the pushbuffer, while still trying to limit the number of jumps